### PR TITLE
feat: add tulipsport support

### PR DESCRIPTION
**Pull request by rdshoep**
online preview: https://tulipsport.github.io/running_page

fix: #252

### CODE REVIEW

**Comment by yihong0618** [comment]
@rdshoep 感谢，我之后看一下哈，这个是郁金香么？

**Comment by rdshoep** [comment]
@yihong0618 对，就是郁金香运动。这个项目非常棒👍

**Comment by yihong0618** [comment]
好的，我没有郁金香账号，我今天注册个实验下哈

File: docs/TulipSport.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+## Github Pages
+### 1.为Github Actions添加代码提交权限
+访问仓库的 `Settings > Actions > General`页面，找到`Workflow permissions`的设置项，将选项配置为`Read and write permissions`，支持CI将运动数据更新后提交到仓库中。
+
```
[the remark this answers is not in the document]
Comment (rdshoep) [reply docs/TulipSport.md]: [配置流程](https://github.com/yihong0618/running_page/pull/379/commits/149029c1d4ff5d901f1a3c5c2212b59fc8ff9e6a#diff-3b6459f9330e366b00c2c21b99b2b154ed014c619559a4ed6d3328585d5995dbR717)已经添加到README-CN.md中

File: scripts/tulipsport_sync.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  location_country = ""
+
+  point_list = detail["map_data_list"]
+  point_list_length = len(point_list)
+  if point_list_length and summary["outdoor"]:
+    first_point = point_list[0]
+    start_latlng = start_point(float(first_point[0]), float(first_point[1]))
+    if point_list_length > 1:
+      last_point = point_list[-1]
+      elapsed_time = datetime.fromisoformat(last_point[6]) - datetime.fromisoformat(first_point[6])
```
[the remark this answers is not in the document]
Comment (rdshoep) [reply scripts/tulipsport_sync.py:101]: I have gave some comments for this~

```python
# 详情接口具体的内容参考文档链接：https://open.tulipsport.com/document
# map_data_list的结构为[ [latitude, longitude, elevation, section, distance, hr, time, cadence], ... ]
```

[commits](https://github.com/yihong0618/running_page/pull/379/commits/149029c1d4ff5d901f1a3c5c2212b59fc8ff9e6a#diff-b31c05526d82abd202a801b07a75b9b6c8f1d6768a0d0a647b03b0e61490da08R92)

File: scripts/tulipsport_sync.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    "location_country": location_country,
+  }
+  return namedtuple("activity_db_instance", activity_db_instance.keys())(*activity_db_instance.values())
+
+def find_last_tulipsport_start_time(track_ids):
+  start_time = None
+  tulipsport_ids = [id for id in track_ids if str(id).startswith(TULIPSPORT_FAKE_ID_PREFIX)]
+  if tulipsport_ids:
+    tulipsport_ids.sort()
+    start_time = datetime.fromtimestamp(int(str(tulipsport_ids[0])[3: -6]), DEFAULT_TIMEZONE)
```
[the remark this answers is not in the document]
Comment (rdshoep) [reply scripts/tulipsport_sync.py]:
```python
# 从模拟的构造ID（特殊前缀(666) + 活动开始时间的timestamp + 活动距离(单位：米，支持单次活动最大距离为999,999米)）中读取时间信息
```

[commits](https://github.com/yihong0618/running_page/pull/379/commits/149029c1d4ff5d901f1a3c5c2212b59fc8ff9e6a#diff-b31c05526d82abd202a801b07a75b9b6c8f1d6768a0d0a647b03b0e61490da08R127)

File: scripts/tulipsport_sync.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+# need to test
+ACTIVITY_LIST_API = "https://open.tulipsport.com/api/v1/feeds4likes?start_time={start_time}&end_time={end_time}"
+ACTIVITY_DETAIL_API = "https://open.tulipsport.com/api/v1/feeddetail?activity_id={activity_id}"
+TULIPSPORT_FAKE_ID_PREFIX = "666"
+
+TIMEZONE_OFFSET = "+08:00"
+TIMEZONE_NAME = "Asia/Shanghai"
+DEFAULT_TIMEZONE = timezone(timedelta(hours=8), TIMEZONE_NAME)
+
+def get_all_activity_summaries(session, headers, start_time=None):
```
[the remark this answers is not in the document]
Comment (rdshoep) [reply scripts/tulipsport_sync.py:30]: Sorry, could you give me some case? I'm the beginner for python, and my IDE named Idea have no error tips.

**Comment by yihong0618** [comment]
@rdshoep can you fix lint using `black`
@ben-29 @shaonianche  can you help me to take a look?

**Comment by rdshoep** [comment]
Yeah, I have reformatted tulipsport_sync.py with default configuration by black@latest. I can't find black's config file in this project. Please help me to check the last commit~

File: docs/TulipSport.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  # please change to your own config.
+  RUN_TYPE: tulipsport
+  ATHLETE: Runner's Name
+  TITLE: Runner Page Title
+  MIN_GRID_DISTANCE: 10 # change min distance here
+  TITLE_GRID: Over 10km Runs # also here
+  GITHUB_NAME: your_github_user_name
+  GITHUB_EMAIL: your_github_email
+```
+3. 在仓库的`Settings > Secrets and variables > Actions`页面添加名为`TULIPSPORT_TOKEN`的新secrets信息，并配置相应token（[如何获取token？](https://tulipsport.rdshoep.com)）
```
Comment (ben-29) [inline docs/TulipSport.md:35]: This section is crucial, and to improve its clarity and conciseness, it would be beneficial to include 'tulip' as a type, just like [Keep](https://github.com/yihong0618/running_page/blob/master/README-CN.md#Keep)

File: scripts/tulipsport_sync.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                        "aid": summary["activity_id"],
+                        "name": "run from tulipsport by " + summary["device"],
+                        "distance": distance,
+                        "moving_time": moving_time,
+                        "elapsed_time": moving_time,
+                        "type": "Run",
+                        "start_date": start_date,
+                        "start_date_local": start_date_local,
+                        "end_date": start_date + moving_time,
+                        "start_date_local": start_date_local + moving_time,
```
Comment (ben-29) [inline scripts/tulipsport_sync.py:71]: remove this line

File: scripts/tulipsport_sync.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        s, headers, find_last_tulipsport_start_time(old_tracks_ids)
+    )
+    activity_summary_list = [
+        activity
+        for activity in activity_summary_list
+        if activity["id"] not in old_tracks_ids
+    ]
+    print(f"{len(activity_summary_list)} new activities to generate")
+    tracks = []
+    old_gpx_ids = os.listdir(GPX_FOLDER)
```
Comment (ben-29) [inline scripts/tulipsport_sync.py:171]: To stay true to the original design, consider add the `--with-gpx` parameter. You can search for this parameter within the codebase to learn more.

**Review by ben-29 — CHANGES_REQUESTED** [review]
Great job! This is a fantastic bridge that will greatly benefit our users. I believe that this integration will allow our users to access COROS/Suunto/Apple & Huawei Health through tulip  as a workaround.

File: README-CN.md [hunk 8]
```diff
@@ -714,6 +714,25 @@ python3(python) scripts/gen_svg.py --from-db --type circular --use-localtime
 
 </details>
 
+## GitHub Pages方案配置
```
[the remark this answers is not in the document]
Comment (rdshoep) [reply README-CN.md]: Lose my eyes😅 I have combined it and remove TulipSport.md.

File: scripts/tulipsport_sync.py (hunk 7 above)
Comment (rdshoep) [reply scripts/tulipsport_sync.py:171]: OK, I will find out later~

Comment (rdshoep) [reply scripts/tulipsport_sync.py:173]: @ben-29 It's done~
https://github.com/yihong0618/running_page/pull/379/commits/be3a9af16e9297269860f03c40b4366c8f9b98a5

File: scripts/tulipsport_sync.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    gpx_track = gpxpy.gpx.GPXTrack()
+    gpx_track.name = track.name
+    gpx.tracks.append(gpx_track)
+
+    # 详情接口具体的内容参考文档链接：https://open.tulipsport.com/document
+    # map_data_list的结构为[ [latitude, longitude, elevation, section, distance, hr, time, cadence], ... ]
+    point_list = detail["map_data_list"]
+    last_section = ""
+    avg_hr = int(detail["avg_hr"])
+    avg_cadence = int(detail["avg_cadence"])
```
Comment (ben-29) [inline scripts/tulipsport_sync.py:205]: I got an error: `Something wrong paring tulipsport id  xxx 'avg_cadence'`
This field may not exist. Could you fix this? @rdshoep

Comment (rdshoep) [reply scripts/tulipsport_sync.py:205]: OK, I try to add a compatible check. My data works fine~

Comment (rdshoep) [reply scripts/tulipsport_sync.py]: I add the compatible check, please help me to check your data.🤝

If the error still occurs, you can tell me your activityId. I find somebody to check the data.

Comment (ben-29) [reply scripts/tulipsport_sync.py]: use `detail.__contains__("avg_cadence")` instead of `detail["avg_cadence"]` for the if statement or it would fail right away.

Comment (rdshoep) [reply scripts/tulipsport_sync.py]: Oh, I see. It's more strict than javascript~ Let me try again😅

Comment (rdshoep) [reply scripts/tulipsport_sync.py]: Done～ I test it by `del detail["avg_cadence"]` manually

**Review by ben-29 — APPROVED** [review]
Well done. Thank you :)

**Comment by rdshoep** [comment]
> 

Thanks for your patience.

**Review by yihong0618 — APPROVED** [review]
LGTM for now, will try use app for one or two runs then merge.

Thanks a lot, really great work

**Comment by yihong0618** [comment]
@ben-29 @rdshoep merging now thanks for your great work and tulipsport

**Comment by rdshoep** [comment]
My pleasure. I'm very glad to contribute to the great project.
Thanks all~ @yihong0618 @ben-29
